### PR TITLE
WIP - Infrakit integration

### PR DESCRIPTION
**Pull request by chungers**
This PR brings infrakit integration + all dependencies.

Specifically, the AWS instance plugin is embedded in this PR to support the use case of creating vm instances of arbitrary images on AWS.  See infrakit/aws-instance.yaml for example:
```

{{ $user := var "user" "string" "username" | prompt "Please enter your user name:" "string" }}
{{ $name := var "name" "string" "name" | prompt "Name?" "string"}}
{{ $imageId := var "ami" "string" "ami" | prompt "AMI?" "string"}}
{{ $instanceType := var "instance-type" "string" "instance type" | prompt "Instance type?" "string"}}
{{ $keyName := var "key" "string" "ssh key name" | prompt "SSH key?" "string"}}
{{ $az := var "az" "string" "availability zone" | prompt "Availability zone?" "string"}}
{{ $subnetId := var "subnet" "string" "subnet id" | prompt "Subnet ID?" "string"}}
{{ $securityGroupId := var "security-group-id" "string" "security group id" | prompt "Security group ID?" "string" }}

Tags:
  infrakit.name: {{ $name }}
  infrakit.created: {{ now | htmlDate }}
  infrakit.user: {{ $user }}

Init: |
  #!/bin/bash
  sudo apt-get update -y
  sudo apt-get install wget curl
  wget -q0- https://get.docker.com | sh

Properties:
  RunInstancesInput:
    BlockDeviceMappings: null
    DisableApiTermination: null
    EbsOptimized: null
    IamInstanceProfile: null
    ImageId: {{ $imageId }}
    InstanceInitiatedShutdownBehavior: null
    InstanceType: {{ $instanceType }}
    KeyName: {{ $keyName }}
    NetworkInterfaces:
    - AssociatePublicIpAddress: true
      DeleteOnTermination: true
      DeviceIndex: 0
      Groups:
      - {{ $securityGroupId }}
      NetworkInterfaceId: null
      SubnetId: {{ $subnetId }}
    Placement:
      Affinity: null
      AvailabilityZone: {{ $az }}
      Tenancy: null
    RamdiskId: null
    SubnetId: null
    UserData: null
  Tags:
    infrakit.name: {{ $name }}
```
 The configuration is a template in YAML format.  Note that there are `var` and `prompt` functions.  The `var` functions can retrieve values from the `--var` or `-v` flags from the command line.  If not specified, the user will be prompted if the `prompt` function is used in the pipeline.   
The command line options are as follows:
```
$ build/testkit create infrakit -h
Create instances using infrakit

Usage:
  build/testkit create infrakit <provider> <url> [flags]

Flags:
      --access-key-id string       IAM access key ID
  -x, --dump                       True to dump to output instead of executing
      --namespace string           Scoping label for resources (default "testkit")
      --region string              AWS region
      --retries int                Number of retries for AWS API operations (default 5)
      --secret-access-key string   IAM access key secret
      --session-token string       AWS STS token
  -v, --var stringSlice            key=value pairs of globally scoped variagbles
  -y, --yaml                       True if input is in yaml format; json is the default

Global Flags:
      --httptest.serve string   if non-empty, httptest.NewServer serves on this address and blocks
      --log int                 log level (default 4)
      --log-caller              include caller function (default true)
      --log-format string       log format: logfmt|term|json (default "term")
      --log-stack               include caller stack
      --log-stdout              log to stdout
```
The command takes a provider (e.g. `aws/ec2-instance`) which will load the AWS instance plugin/driver.  The URL is the location of the config input YAML (eg. one shown above).  The url can be from local file (file://) or from a server (http:// https://)   Template files can source other files to maximize reuse and infrakit will 'crawl' all the files specified until a full config is rendered.

When the config template is rendered in full, it becomes the input to the infrakit instance plugin and the instance plugin will create 1 instance.  Multiple instances require different calls.  It's also impossible to have N instances, but it's a more complex template and will be added later.

### CODE REVIEW

File: infrakit/aws-instance.yaml [hunk 1]
```diff
@@ -0,0 +1,48 @@
+
+{{ $user := var "user" "string" "username" | prompt "Please enter your user name:" "string" }}
+{{ $name := var "name" "string" "name" | prompt "Name?" "string"}}
+{{ $imageId := var "ami" "string" "ami" | prompt "AMI?" "string"}}
+{{ $instanceType := var "instance-type" "string" "instance type" | prompt "Instance type?" "string"}}
+{{ $keyName := var "key" "string" "ssh key name" | prompt "SSH key?" "string"}}
+{{ $az := var "az" "string" "availability zone" | prompt "Availability zone?" "string"}}
+{{ $subnetId := var "subnet" "string" "subnet id" | prompt "Subnet ID?" "string"}}
+{{ $securityGroupId := var "security-group-id" "string" "security group id" | prompt "Security group ID?" "string" }}
```
Comment (nwt) [inline infrakit/aws-instance.yaml]: @chungers: Should these be `flag` instead of `var`?

**Comment by vdemeester** [comment]
@chungers Some of the flags are aws specific, what is the reason ? 👼

File: infrakit/aws-instance.yaml (hunk 1 above)
Comment (chungers) [inline infrakit/aws-instance.yaml]: Not using the dynamic CLI bits in infrakit here so it's just simple `var` and settable via `--var` flag in the CLI.  This just means these are not unique CLI flags visible in help (`-h`).

**Comment by chungers** [comment]
@vdemeester  - just a placeholder as is the `infrakit` subcommand right now....   I personally think config files / yaml is better than CLI flags or environment variables.  But that's not settled yet (see @dhiltgen Docker Machine integration).  So this is a placeholder.

File: vendor.conf [hunk 2]
```diff
@@ -8,7 +8,7 @@ github.com/aokoli/goutils                               1.0.0
 github.com/armon/go-radix                               4239b77
 github.com/aws/aws-sdk-go                               v1.4.20
 github.com/davecgh/go-spew                              v1.0.0-3-g6d21280
-github.com/docker/infrakit                              05d22a5
```
Comment (vdemeester) [inline vendor.conf]: Shouldn't this be in the previous commit ? 👼 (sorry if it's a nit 😛)

File: testkit/cmd/create.go [hunk 3]
```diff
@@ -6,55 +6,110 @@ import (
...
+
+			instancePlugin, err := infrakit.GetInstancePlugin(provider)
+			if err != nil {
+				return err
+			}
+			log.Info("infrakit found plugin", "provider", provider, "url", templateURL, "instance", instancePlugin)
+
+			view, err := infrakit.ReadFromStdinIfElse(
+				func() bool { return args[0] == "-" },
+				func() (string, error) { return provisionProcessTemplate(args[0]) },
```
Comment (nwt) [inline testkit/cmd/create.go]: @chungers: `args[0]` is the provider, so these two funcs should use `templateURL` instead.

File: infrakit/aws-instance.yaml (hunk 1 above)
Comment (nwt) [inline infrakit/aws-instance.yaml]: In that case, I think these need to look like `{{ $blah := var "blah" }}` because they cause `prompt` to panic.

**Comment by chungers** [comment]
I am closing this PR and let's continue on #59

**Pull request closed without merging**